libs/core/libs/shared/src/document_repo.rs

```diff
@@ -1,15 +1,156 @@
+            DocEvent::Write(_, timestamp) => timestamp,
+        }
+    }
+    pub fn id(&self) -> Uuid {
+        match *self {
+            DocEvent::Read(id, _) => id,
+            DocEvent::Write(id, _) => id,
+        }
+    }
+}
+
+#[derive(Default, Copy, Clone, PartialEq)]
+pub struct StatisticValue {
+    pub raw: i64,
+    pub normalized: f64,
+}
+impl StatisticValue {
+    fn from_raw(raw: i64) -> StatisticValue {
+        StatisticValue { raw, normalized: f64::default() }
+    }
+}
+impl Ord for StatisticValue {
+    fn cmp(&self, other: &Self) -> Ordering {
+        (self.raw).cmp(&other.raw)
+    }
+}
+
+impl PartialOrd for StatisticValue {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Eq for StatisticValue {}
+
+#[derive(Clone, Copy)]
+pub struct StatisticValueRange {
+    pub max: StatisticValue,
+    pub min: StatisticValue,
+}
+impl StatisticValue {
+    pub fn normalize(&mut self, range: StatisticValueRange) {
+        self.normalized = (self.raw - range.min.raw) as f64 / (range.max.raw - range.min.raw) as f64
+    }
+}
+/// DocActivityMetrics stores key document activity features, which are used to recommend relevant documents to the user.
+/// Here's a walkthrough of the recommendation procedure: collect 1k most recent document events (write/read), use that activity to construct a DocActivtyMetrics struct for each document. Min-max normalizes the activity features, then rank the documents.
+/// latest_read_timestamp: the latest epoch timestamp that the user read a document
+/// latest_write_timestamp: the latest epoch timestamp that the user wrote a document
+/// the total number of times that a user reads a document
+/// the total number of times that a user wrote a document
+#[derive(Default, Copy, Clone)]
+pub struct DocActivityMetrics {
+    pub id: Uuid,
+    pub latest_read_timestamp: StatisticValue,
+    pub latest_write_timestamp: StatisticValue,
+    pub read_count: StatisticValue,
+    pub write_count: StatisticValue,
+}
+impl DocActivityMetrics {
+    pub fn rank(&self) -> i64 {
+        (self.latest_read_timestamp.normalized + self.latest_write_timestamp.normalized) as i64 * 70
+            + (self.read_count.normalized + self.write_count.normalized) as i64 * 30
+    }
+}
+pub trait Stats {
+    fn get_activity_metrics(self) -> Vec<DocActivityMetrics>;
+}
+impl<'a, T> Stats for T
+where
+    T: Iterator<Item = &'a DocEvent>,
+{
+    fn get_activity_metrics(self) -> Vec<DocActivityMetrics> {
+        let mut result = Vec::new();
+
+        let mut set = HashMap::new();
+        for event in self {
+            match set.get_mut(&event.id()) {
+                None => {
+                    set.insert(event.id(), vec![event]);
+                }
+                Some(events) => {
+                    events.push(event);
+                }
+            }
+        }
+
+        for pair in set {
+            let read_events: Vec<&&DocEvent> = pair
```

Comment:
I don't think you actually need random access to the read events, you just need either the `.count()` or the `.last()` element. `collect()`'ing into a `vec` is an expensive allocation that it doesn't look like you benefit from here.

Reply:
it's either collecting into a vec or cloning the iter.